### PR TITLE
KAFKA-17840: Move ReplicationQuotaManager, ClientRequestQuotaManager and QuotaFactory to server module

### PR DESCRIPTION
**Pull request by FrankYang0529**
Convert `ReplicationQuotaManager`, `ClientRequestQuotaManager`, and `QuotaFactory` to Java.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

File: core/src/main/java/kafka/server/QuotaFactory.java [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+
+import java.util.Optional;
+
+import scala.Option;
+
+public class QuotaFactory {
+
+    public static final UnboundedQuota UNBOUNDED_QUOTA = new UnboundedQuota();
+
+    public static class UnboundedQuota implements ReplicaQuota {
```
Comment (chia7712) [inline core/src/main/java/kafka/server/QuotaFactory.java]:
```java
    public static final ReplicaQuota UNBOUNDED_QUOTA = new ReplicaQuota() {
        @Override
        public boolean isThrottled(TopicPartition topicPartition) {
            return false;
        }

        @Override
        public boolean isQuotaExceeded() {
            return false;
        }

        @Override
        public void record(long value) {
            // No-op
        }
    };
```
WDYT?

File: core/src/main/java/kafka/server/ClientRequestQuotaManager.java [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+
+    /**
+     * Records that a user/clientId changed request processing time being throttled. If quota has been violated, return
+     * throttle time in milliseconds. Throttle time calculation may be overridden by sub-classes.
+     * @param request client request
+     * @return Number of milliseconds to throttle in case of quota violation. Zero otherwise
+     */
+    public int maybeRecordAndGetThrottleTimeMs(RequestChannel.Request request, long timeMs) {
+        if (quotasEnabled()) {
+            request.setRecordNetworkThreadTimeCallback(timeNanos -> {
```
Comment (chia7712) [inline core/src/main/java/kafka/server/ClientRequestQuotaManager.java:76]: Could you please change the input type of `setRecordNetworkThreadTimeCallback` from `Long => Unit` to `java.util.function.Consumer`? Using the scala type in java is a bit weird

File: core/src/main/java/kafka/server/ReplicationQuotaManager.java [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+            lock.writeLock().unlock();
+        }
+    }
+
+    /**
+     * Check if the quota is currently exceeded
+     */
+    @Override
+    public boolean isQuotaExceeded() {
+        try {
```
Comment (chia7712) [inline core/src/main/java/kafka/server/ReplicationQuotaManager.java:84]:
```java
        try {
            sensor().checkQuotas();
            return false;
        } catch (QuotaViolationException qve) {
            LOGGER.trace("{}: Quota violated for sensor ({}), metric: ({}), metric-value: ({}), bound: ({})",
                replicationType, sensor().name(), qve.metric().metricName(), qve.value(), qve.bound());
            return true;
        }
```

**Review by chia7712 — COMMENTED** [review]
@FrankYang0529 thanks for this patch!

**Review by chia7712 — APPROVED** [review]

**Comment by mimaison** [comment]
This does not move ReplicationQuotaManager, ClientRequestQuotaManager and QuotaFactory to the server module. These classes are still in core, they were just rewritten in Java.

We probably need to reopen https://issues.apache.org/jira/browse/KAFKA-17840 to actually move them? or am I missing something?

**Comment by chia7712** [comment]
> These classes are still in core, they were just rewritten in Java.

Oops, looks like a mistake. Will handle it later

**Comment by mimaison** [comment]
No worries. I reopened https://issues.apache.org/jira/browse/KAFKA-17840. It seems we can't quite move these classes yet, `ClientRequestQuotaManager` still depends on `kafka.network.RequestChannel`. We're pretty much at the point where we're blocked by `KafkaConfig` for all the remaining classes.

**Comment by FrankYang0529** [comment]
Sorry for that. I reassign `KafkaConfig` (https://issues.apache.org/jira/browse/KAFKA-15853) to @chia7712.  We will handle it together.

**Pull request merged**